### PR TITLE
feat: add fields for enhancing global index

### PR DESCRIPTION
**Pull request by L-maple**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/41991
Issue Number: ref https://github.com/pingcap/tidb/issues/42065

related PR: https://github.com/pingcap/tidb/pull/41992?w=1

Problem Summary:

### What is changed and how it works?

Proposal: [global index](https://github.com/pingcap/tidb/blob/master/docs/design/2020-08-04-global-index.md) 

What's Changed:

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### CODE REVIEW

File: proto/executor.proto [hunk 1]
```diff
@@ -154,6 +154,8 @@ message IndexScan {
 	optional bool desc = 4 [(gogoproto.nullable) = false];
 	optional bool unique = 5; // check whether it is a unique index.
 	repeated int64 primary_column_ids = 6;
+	optional bool is_global = 7; // check whether it is a global index.
```
Comment (Defined2014) [inline proto/executor.proto:157]: Can you briefly explain why these two parameters are added?

And I think you should also modify its related logic on the TiKV side.

Comment (L-maple) [reply proto/executor.proto:157]: `is_global` means whether it is a global index. Only when `is_global` is true, we should consider `partition_ids`. TiKV then filters out the rows by comparing `partition_ids` and PartitionID in index value.

For global index, Not all indexes' row should be returned when partitions are specified or some partitions are in reorg.
e.g., select * from pt partition(p0) use index(global_idx); The IndexScan DAG should contain p0 to filter the needed partitions' rows. 
e.g., pt has p0,p1,p2 partitions. When p0 is in reorg, its index rows should not be queried. 

In summary, necessary info should be contained in the IndexScan, so that TiKV can handle these circumstances correctly.

Comment (L-maple) [reply proto/executor.proto:157]: OK, I will propose the PR for TiKV, and mention the PR here. 

But please review this PR first, so that this tipb PR can be used in TiDB PR. Then the TiDB PR can be tested.

Comment (mjonss) [reply proto/executor.proto:157]: Is it possible to use a Selection instead of combining an Index Scan with a filter/selection? I.e. do we need to add this PR?

Comment (L-maple) [reply proto/executor.proto:157]: @mjonss  Aha, it's a really nice idea. But it is really difficult for me in two aspects: 
1) What conditions I should put into Selection. Because ExtraPartitionID is not a real column.
2) When I should construct the Selection.

**Review by Defined2014 — APPROVED** [review]

File: proto/executor.proto (hunk 1 above)
Comment (mjonss) [reply proto/executor.proto:157]: @L-maple I don't know that area of the code very well, but @windtalker have some insights?

In my view the ExtraPartitionID should be treated as a column when communicating, just that it has no connection to the table's columns, it can still be seen as a normal column when reading the global index in TiKV/unistore.

Some references when reading from TiKV or unistore in https://github.com/tikv/tikv/pull/11931 and there are some interesting comments in https://github.com/tikv/tikv/issues/11888.

Comment (mjonss) [reply proto/executor.proto:157]: I'm OK in doing the filtering in TiDB for now and postpone the optimization of pushing down the filtering to a future PR. I just don't think we should change tipb repo like this, since I think the current protocol already have a more generic way of solving this.

Comment (L-maple) [reply proto/executor.proto:157]: @mjonss @Defined2014 OK, I agree postponing the optimization.
I also don't know well about the conditions handling. And I believe if ExtraTableID is handled well, ExtraPartitionID can also. 

I will propose the optimized version in the future. This PR can be closed.

Comment (L-maple) [reply proto/executor.proto:157]: @mjonss  I remove the push down logic, but remain related unit tests. This optimization(or fix) can be realized in future.

Comment (windtalker) [reply proto/executor.proto:157]: I aggree that `ExtraPartitionID` should be treated as a normal column, and I think it is not a big problem about when/how to construct the selection, you can always contruct the filter/selection when your current code is setting `is_global_index` and `partition_ids`. There might be one benifit by putting `is_global_index/partition_ids` inside index scan: the index scan itself can skip some blocks based on these information. But for a well design compiler, it should support push down filter to scan(not sure if TiKV support this)

**Comment by L-maple** [comment]
@Defined2014  Hi, how could I close this PR? `/close` seems not function.

**Comment by Defined2014** [comment]
Do you have permission to press the close pull button? 
image

**Comment by L-maple** [comment]
> Do you have permission to press the close pull button? image

Oh, I find it! Thanks~
